libraries/chef_updater.rb

```diff
@@ -107,10 +114,29 @@ def action_run
             block { throw :end_client_run_early_due_to_chef_upgrade }
             action :nothing
           end
+
+          if platform?('solaris2') && new_resource.use_ips_package
+            # Remove it from the old packaging system if it's there
+            # Otherwise, you end up with both
+            # The solaris_package resource does not remove packages properly.
+            # https://github.com/chef/chef/blob/master/lib/chef/provider/package/solaris.rb#L59
+            # it uses pkginfo -d which tells you about the package *in a source*
+            # which is not the package *installed on the system*
+            # load_current_resource is called when you attempt to use action :remove
+            # Ergo, we do it the evil way
+            # The "yes" is necessary because pkgrm wants to be interactive
+            # and we haven't set up the "admin file" to make it accept -n for chef.
+            bash 'uninstall old chef' do
+              code 'yes | pkgrm chef'
+              only_if 'pkginfo -l chef'
+            end
+          end
+
           package new_resource.package_name do
             action :upgrade
             provider Chef::Provider::Package::Dpkg if platform?('ubuntu')
-            provider Chef::Provider::Package::Solaris if platform?('solaris2')
+            provider Chef::Provider::Package::Ips if platform?('solaris2') && new_resource.use_ips_package
```

Comment:
wait, what are you wanting to change here? it looks like part of your comment got wiped.